tools/dtest/harness/harness.go

```diff
@@ -0,0 +1,305 @@
+package harness
+
+import (
+	// pprof import
+	_ "net/http/pprof"
+
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"sync/atomic"
+	"time"
+
+	"github.com/m3db/m3db/tools/dtest/config"
+	"github.com/m3db/m3db/x/m3em/convert"
+	m3dbnode "github.com/m3db/m3db/x/m3em/node"
+
+	etcdclient "github.com/m3db/m3cluster/client/etcd"
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3cluster/services/placement"
+	"github.com/m3db/m3cluster/shard"
+	"github.com/m3db/m3em/build"
+	"github.com/m3db/m3em/cluster"
+	hb "github.com/m3db/m3em/generated/proto/heartbeat"
+	"github.com/m3db/m3em/node"
+	xerrors "github.com/m3db/m3x/errors"
+	"github.com/m3db/m3x/instrument"
+	"github.com/m3db/m3x/log"
+	xtcp "github.com/m3db/m3x/tcp"
+	"google.golang.org/grpc"
+)
+
+type closeFn func() error
+
+// DTestHarness makes it easier to maintain the lifecycle of
+// resources used for DTests.
+type DTestHarness struct {
+	io.Closer
+
+	closing          int32
+	closers          []closeFn
+	cliOpts          *config.CLIOpts
+	conf             *config.Configuration
+	iopts            instrument.Options
+	logger           xlog.Logger
+	placementService services.PlacementService
+	nodeOpts         node.Options
+	clusterOpts      cluster.Options
+	nodes            []m3dbnode.Node
+}
+
+// New constructs a new DTestHarness
+func New(cliOpts *config.CLIOpts, logger xlog.Logger) *DTestHarness {
+	dt := &DTestHarness{
+		cliOpts: cliOpts,
+		logger:  logger,
+		iopts:   instrument.NewOptions().SetLogger(logger),
+	}
+
+	conf, err := config.New(cliOpts.M3EMConfigPath)
+	if err != nil {
+		logger.Fatalf("unable to read configuration file: %v", err.Error())
+	}
+	dt.conf = conf
+	dt.startPProfServer()
+
+	pSvc, err := placementService(dt.m3dbServiceID(),
+		conf.KV.NewOptions(), defaultPlacementOptions(dt.iopts))
+	if err != nil {
+		logger.Fatalf("unable to create placement service %v", err)
+	}
+	dt.placementService = pSvc
+
+	no := conf.M3EM.Node.Options(dt.iopts)
+	dt.nodeOpts = no.SetHeartbeatOptions(
+		no.HeartbeatOptions().
+			SetEnabled(true).
+			SetHeartbeatRouter(dt.newHeartbeatRouter()))
+
+	nodes, err := dt.conf.M3EM.M3DBNodes(dt.nodeOpts, cliOpts.NumNodes)
+	if err != nil {
+		logger.Fatalf("unable to create m3db nodes: %v", err)
+	}
+	dt.nodes = nodes
+
+	dt.clusterOpts = cluster.NewOptions(pSvc, dt.iopts).
+		SetServiceBuild(newBuild(logger, cliOpts.M3DBBuildPath)).
+		SetServiceConfig(newConfig(logger, cliOpts.M3DBConfigPath)).
+		SetSessionToken(cliOpts.SessionToken).
+		SetSessionOverride(cliOpts.SessionOverride).
+		SetNumShards(conf.DTest.NumShards)
+
+	if cliOpts.InitialReset {
+		svcNodes, err := convert.AsServiceNodes(nodes)
+		if err != nil {
+			logger.Fatalf("unable to cast nodes: %v", err)
+		}
+		exec := node.NewConcurrentExecutor(svcNodes, dt.clusterOpts.NodeConcurrency(), dt.clusterOpts.NodeOperationTimeout(), func(n node.ServiceNode) error { return n.Teardown() })
```

Comment:
nit: maybe make this line a little shorter?  I'm not usually one to be barking up the "only use 80 char width" but this one is pretty long.

I have a tendency to:
```go
concurrency := dt.clusterOpts.NodeConcurrency()
timeout := dt.clusterOpts.NodeOperationTimeout()
teardownFn := func(n node.ServiceNode) error { return n.Teardown() }
exec := node.NewConcurrentExecutor(svcNodes, concurrency, timeout, teardownFn)
```